torch/csrc/distributed/c10d/reducer.cpp

```diff
@@ -572,20 +609,56 @@ void Reducer::prepare_for_backward(
 
 // A bucket with one or more dense tensors needs to be unflattened.
 void Reducer::finalize_bucket_dense(Bucket& bucket) {
-  for (auto& replica : bucket.replicas) {
```

Comment:
Shall we add an `TORCH_INTERNAL_ASSERT`/`AT_ERROR` that `local_used_maps_reduced_ ` must be false when entering this method.

Reply:
Done with cc28ee6.